File: examples/item-list/src/main/java/com/gurunars/item_list/example/AnimalItem.java
package com.gurunars.item_list.example;


import com.gurunars.item_list.Item;

class AnimalItem extends Item<Integer> {

    enum Type {
        MONKEY, TIGER, WOLF, LION
    }

    public void update() {
        this.setPayload(getPayload()+1);
    }

    public AnimalItem(long id, Type type) {
        super(type, id, 0);
    }

    @Override
    public String toString() {
        return "" + getId() + " @ " + getPayload();
    }
}

File: examples/item-list/src/main/kotlin/com/gurunars/item_list/example/AnimalItem.kt
package com.gurunars.item_list.example


import com.gurunars.item_list.Item

internal class AnimalItem(override val id: Long, override val type: AnimalItem.Type) : Item {

    enum class Type {
        MONKEY, TIGER, WOLF, LION
    }

    private var version: Int = 0

    fun update() {
        this.version++
    }

    override fun equals(obj: Any?): Boolean {
        if (obj !is AnimalItem) {
            return false
        }
        return id == obj.id && type == obj.type && version == obj.version
    }

    override fun toString(): String {
        return "" + id + " @ " + version
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + version
        return result
    }
}
